pubsub/fakepubsub/fake.go

```diff
@@ -0,0 +1,209 @@
+// Copyright 2018 The Go Cloud Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package fakepubsub provides an in-memory fake pubsub implementation.
+// This should not be used for production: it is intended for local
+// development.
+//
+// fakepubsub does not support any types for As.
+package fakepubsub
```

Comment:
I like `rampubsub`. It's nice and short.

Reply:
Fake is absolutely the right term. See https://testing.googleblog.com/2013/05/testing-on-toilet-dont-overuse-mocks.html, search for "fake". The post doesn't go into detail, but a _fake_ is a test double that simulates the real behavior of the system in a simplified way.

`rampubsub` might give someone the impression that this is a production-quality in-memory pubsub system, rather than something that is only usable for testing.

Reply:
Makes sense.

Reply:
Should we call `fileblob` `fakeblob` instead? It is also a simplified version of a Cloud blob storage system, in that it is limited to a single machine rather than global namespace.

If the driver satisfies the interface, is it fake? Similar to `fileblob`, the main difference is that it is limited to a single process, unlike most "real" implementations.

I can also imagine other kinds of fakes, e.g. a topic where you give it a bunch of messages at construction time and it sends them out to subscribers on some schedule.

@zombiezen any thoughts on naming?

Reply:
`fileblob` makes sense as-is: we had a small debate along these lines there and naming it descriptively of how it works seems appropriate. `mempubsub` seems like the right name to me. If a thing is primarily useful for testing, it will be reflected in how people use it. If it has other uses, that will be reflected too.

Reply:
If we rename it to `mempubsub` can we the package doc say it is "intended primarily for testing"?

Reply:
Sounds good.

Reply:
SGTM. Maybe "local development and testing"?